File: scripts/communicator.js
var Communicator = function (win, self) {
  this.dest = win;
  this._callback_maps = {};
  this._action_handlers = {};
  this.source = self || window;
  this.source.addEventListener('message', this._onmessage.bind(this), false);
};

Communicator.prototype._postMessage = function (dest, action, value, callback, reply_to) {
  var reqid = null;
  
  if (callback) {
    var map = this._callback_maps[action];
    if (!map) {
      map = this._callback_maps[action] = {};
    }
    do {
      reqid = Math.random();
    } while (map[reqid]);
    map[reqid] = callback;
  }
  
  dest.postMessage({
    'action': action,
    'reqid': reqid,
    'reply_to': reply_to,
    'value': value
  }, '*');
}

Communicator.prototype.createResponder = function (e) {
  var that = this;
  return function (value, callback) {
    that._postMessage(e.source, e.data.action, value, callback, e.data.reqid)
  };
};

Communicator.prototype._onmessage = function (e) {
  if (e.data.reply_to) {
    var map = this._callback_maps[e.data.action];
    if (map) {
      var callback = map[e.data.reply_to];
      if (callback) {
        delete map[e.data.reply_to];
        callback(e.data.value, this.createResponder(e));
      }
    }
  } else {
    var callbacks = this._action_handlers[e.data.action];
    if (callbacks) {
      var responder = this.createResponder(e);
      callbacks.forEach(function (cb) {
        cb(e.data.value, responder);
      });
    }
  }
};

Communicator.prototype.send = function (action, value, callback) {
  this._postMessage(this.dest, action, value, callback);
  
  return this;
};

// Alt signature: .on({ 'action1': callback1, 'action2': callback2 });
Communicator.prototype.on = function (action, callback) {
  if (typeof action === 'object') {
    for (var a in action) {
      if (action.hasOwnProperty(a)) {
        this.on(a, action[a])
      }
    }
  } else {
    var callbacks = this._action_handlers[action];
    if (!callbacks) {
      callbacks = this._action_handlers[action] = [];
    }
    callbacks.push(callback);
  }
  
  return this;
};

File: scripts/communicator.dart
/*!
 * Copyright (C) 2012, The SwitchyOmega Authors. Please see the AUTHORS file
 * for details.
 *
 * This file is part of SwitchyOmega.
 *
 * SwitchyOmega is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * SwitchyOmega is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with SwitchyOmega.  If not, see <http://www.gnu.org/licenses/>.
 */

#library('communicator');

#import('dart:html');

typedef void CommunicatorCallback(Object value, [Function respond]);

class Communicator {
  Window dest;
  Window source;
  Map<String, Map<String, CommunicatorCallback>> _callback_maps;
  Map<String, List<CommunicatorCallback>> _action_handlers;
  
  Communicator([this.dest = null, this.source = null]) {
    if (this.source == null) {
      this.source = window;
    }
    _callback_maps = new Map<String, Map<String, CommunicatorCallback>>();
    _action_handlers = new Map<String, List<CommunicatorCallback>>();
    this.source.on.message.add(this._onmessage);
  }
  
  void _postMessage(Window destWin, String action, Object value,
                    CommunicatorCallback callback, [String reply_to = null]) {
    String reqid;
    
    if (callback != null) {
      var map = this._callback_maps[action];
      if (map == null) {
        map = this._callback_maps[action] = new Map<String, CommunicatorCallback>();
      }
      do {
        reqid = Math.random().toString();
      } while (map[reqid] != null);
      map[reqid] = callback;
    }
    
    destWin.postMessage({
      'action': action,
      'reqid': reqid,
      'reply_to': reply_to,
      'value': value
    }, '*');
  }
  
  CommunicatorCallback createResponder(MessageEvent e) {
    return (Object value, [Function respond]) {
      Map<String, Object> data = e.data; // CAST
      this._postMessage(e.source, data['action'], value, respond, data['reqid']);
    };
  }
  
  void _onmessage(MessageEvent e) {
    Map<String, Object> data = e.data; // CAST
    var reply_to = data['reply_to'];
    if (reply_to != null) {
      var map = this._callback_maps[data['action']];
      if (map != null) {
        var callback = map[reply_to];
        if (callback != null) {
          map.remove(reply_to);
          callback(data['value'], this.createResponder(e));
        }
      }
    } else {
      var callbacks = this._action_handlers[data['action']];
      if (callbacks != null) {
        var responder = this.createResponder(e);
        callbacks.forEach((cb) {
          cb(data['value'], responder);
        });
      }
    }
  }
  
  Communicator send(String action, [Object value, CommunicatorCallback callback]) {
    this._postMessage(this.dest, action, value, callback);
    
    return this;
  }
  
  Communicator on(action, CommunicatorCallback callback) {
    if (action is String) {
      var actionString = action; // CAST
      var callbacks = this._action_handlers[action];
      if (callbacks == null) {
        callbacks = this._action_handlers[action] = [];
      }
      callbacks.push(callback);
    } else {
      Map<String, CommunicatorCallback> actionMap = action; // CAST
      actionMap.forEach((a, cb) {
        this.on(a, cb);
      });
    }
    
    return this;
  }
}
